### DISCUSSION TITLE
Extension methods for MockBuilder

### DISCUSSION THREAD

**Discussion by AE1NS** [help-me-mom/ng-mocks#5404] [Q&A] [answered]
I would love to write some reusable extension methods for the MockBuilder. But as MockBuilder is a function and not a class, I'm struggling a bit how to achieve this.
Thought I could use something like this:
.d.ts:
```
export {};

declare module 'ng-mocks' {
    interface IMockBuilder {
        mockCustom(): IMockBuilder;
    }
}

```
.ts:
```
MockBuilder.prototype.mockCustom = function(): IMockBuilder {
    // some custom stuff on the MockBuilder
    return this;
};
```
test.spec.ts:
```
beforeEach(() =>
    MockBuilder(TestComponent)
        .mockCustom()
);
```
The type is recognized correctly but it will not run, as 'MockBuilder.prototype' cant work on a function. Can you tell me if can extend it on another way externally?

**Comment by AE1NS** [comment]
Ok, now I got it. The function MockBuilder returns a MockBuilderPerformance which extends MockBuilderPromise which implements IMockBuilder. But as IMockBuilder is an interface, I have to extend the prototype of MockBuilderPerformance or MockBuilderPromise. Both of them are not exported, so its not possible for me. Do you know another way or do you think its impossible to do this with external code?

I think the only way would be a a CustomMockBuilder that creates a MockBuilder instance, extends the variable and returns it.

**Reply by satanTime** [reply]
Hey there,

it sounds like a feature request: https://github.com/help-me-mom/ng-mocks/issues/5417

could you add there a couple of ideas how you would like to extend `MockBuilder`?

**Reply by satanTime** [reply, marked as answer]
He solution is described here: https://ng-mocks.sudo.eu/api/MockBuilder#extending-mockbuilder
